File: parser/http.go
package parser

import (
	"bufio"
	"bytes"
	"net/http"
	"strconv"
	"strings"
)

// parseHTTP does a best-effort layer-7 parse of TCP payload for HTTP/1.x.
func parseHTTP(p *Packet) {
	if p.Protocol != "TCP" || len(p.Payload) < 4 {
		return
	}

	// Try parsing as HTTP request
	if isHTTPRequest(p.Payload) {
		req, err := http.ReadRequest(bufio.NewReader(bytes.NewReader(p.Payload)))
		if err == nil {
			p.HTTPMethod = req.Method
			p.HTTPHost = req.Host
			p.HTTPPath = req.URL.RequestURI()
			return
		}
	}

	// Try parsing as HTTP response
	if isHTTPResponse(p.Payload) {
		resp, err := http.ReadResponse(bufio.NewReader(bytes.NewReader(p.Payload)), nil)
		if err == nil {
			p.HTTPStatus = resp.StatusCode
		}
	}
}

func isHTTPRequest(b []byte) bool {
	methods := []string{"GET ", "POST ", "PUT ", "DELETE ", "HEAD ", "OPTIONS ", "PATCH ", "CONNECT "}
	for _, m := range methods {
		if bytes.HasPrefix(b, []byte(m)) {
			return true
		}
	}
	return false
}

func isHTTPResponse(b []byte) bool {
	if !bytes.HasPrefix(b, []byte("HTTP/")) {
		return false
	}
	line := strings.SplitN(string(b[:min(len(b), 20)]), " ", 3)
	if len(line) < 2 {
		return false
	}
	_, err := strconv.Atoi(line[1])
	return err == nil
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

File: parser/http.py
from parser.packet import Packet

_METHODS = (b"GET ", b"POST ", b"PUT ", b"DELETE ", b"HEAD ", b"OPTIONS ", b"PATCH ", b"CONNECT ")


def parse_http(p: Packet) -> None:
    if p.protocol != "TCP" or len(p.payload) < 4:
        return

    payload = p.payload

    if any(payload.startswith(m) for m in _METHODS):
        _parse_request(p, payload)
    elif payload.startswith(b"HTTP/"):
        _parse_response(p, payload)


def _parse_request(p: Packet, payload: bytes) -> None:
    try:
        first_line_end = payload.index(b"\r\n")
        first_line = payload[:first_line_end].decode("utf-8", errors="replace")
        parts = first_line.split(" ", 2)
        if len(parts) >= 2:
            p.http_method = parts[0]
            p.http_path = parts[1]

        headers = payload[first_line_end + 2:]
        for line in headers.split(b"\r\n"):
            if line.lower().startswith(b"host:"):
                p.http_host = line[5:].strip().decode("utf-8", errors="replace")
                break
    except (ValueError, UnicodeDecodeError):
        pass


def _parse_response(p: Packet, payload: bytes) -> None:
    try:
        first_line_end = payload.index(b"\r\n")
        first_line = payload[:first_line_end].decode("utf-8", errors="replace")
        parts = first_line.split(" ", 2)
        if len(parts) >= 2:
            p.http_status = int(parts[1])
    except (ValueError, UnicodeDecodeError):
        pass
